netket/operator/local_values.py

```diff
@@ -1,45 +1,114 @@
+def Heisenberg(hilbert, J=1, sign_rule=None):
+    """
+    Constructs a new ``Heisenberg`` given a hilbert space.
+
+    Args:
+        hilbert: Hilbert space the operator acts on.
+        J: The strength of the coupling. Default is 1.
+        sign_rule: If enabled, Marshal's sign rule will be used. On a bipartite
+                   lattice, this corresponds to a basis change flipping the Sz direction
+                   at every odd site of the lattice. For non-bipartite lattices, the
+                   sign rule cannot be applied. Defaults to True if the lattice is
+                   bipartite, False otherwise.
+
+    Examples:
+     Constructs a ``Heisenberg`` operator for a 1D system.
+
+        >>> import netket as nk
+        >>> g = nk.graph.Hypercube(length=20, n_dim=1, pbc=True)
+        >>> hi = nk.hilbert.Spin(s=0.5, total_sz=0, graph=g)
+        >>> op = nk.operator.Heisenberg(hilbert=hi)
+        >>> print(op.hilbert.size)
+        20
+    """
+    if sign_rule is None:
+        sign_rule = hilbert.graph.is_bipartite
+
+    sz_sz = _np.array([[1, 0, 0, 0], [0, -1, 0, 0],
+                       [0, 0, -1, 0], [0, 0, 0, 1]])
+    exchange = _np.array(
+        [[0, 0, 0, 0], [0, 0, 2, 0], [0, 2, 0, 0], [0, 0, 0, 0]])
+    if sign_rule:
+        if not hilbert.graph.is_bipartite:
+            raise ValueError(
+                "sign_rule=True specified for a non-bipartite lattice")
+        heis_term = sz_sz - exchange
+    else:
+        heis_term = sz_sz + exchange
+    return GraphOperator(hilbert, bondops=[J * heis_term])
+
+
+@jit(nopython=True)
+def _local_values_kernel(log_vals, log_val_primes, mels, sections, out):
+    low_range = 0
+    for i, s in enumerate(sections):
+        out[i] = (mels[low_range:s] *
+                  _np.exp(log_val_primes[low_range:s] - log_vals[i])).sum()
+        low_range = s
+
+
+def _local_values_impl(op, machine, v, log_vals, out):
+
+    sections = _np.empty(v.shape[0], dtype=_np.int32)
+    v_primes, mels = op.get_conn_flattened(v, sections)
+
+    log_val_primes = machine.log_val(v_primes)
+
+    _local_values_kernel(log_vals, log_val_primes, mels, sections, out)
 
 
 def _local_values_op_op_impl(op, machine, v, log_vals, out):
 
     vprimes, mels = op.get_conn(v)
 
-    vold = [_np.tile(vi, (vprime.shape[0], 1)) for (vi, vprime) in zip(v, vprimes)]
+    vold = [_np.tile(vi, (vprime.shape[0], 1))
+            for (vi, vprime) in zip(v, vprimes)]
 
-    log_val_primes = [machine.log_val(vprime, v) for (vprime, v) in zip(vprimes, vold)]
+    log_val_primes = [machine.log_val(vprime, v)
+                      for (vprime, v) in zip(vprimes, vold)]
 
-    _local_values_kernel(log_vals, log_val_primes, mels, out)
-
-    # for k, sample in enumerate(v):
-    #
-    #     lvd = machine.log_val(vprimes[k], v)
-    #
-    #     out[k] = (mels[k] * _np.exp(lvd - log_vals[k])).sum()
+    _c_local_values_kernel(log_vals, log_val_primes, mels, out)
```

Comment:
I think you can already get rid of the c local values kernel.

Just replace with 
```python
    sections = _np.empty(v.shape[0], dtype=_np.int32)
    _local_values_kernel(log_vals, log_val_primes, mels, sections, out)
```

and I think it should work

Reply:
The custom `op_op` logic is the one to compute log_val_primes